src/nnet3/nnet-simple-component.cc

```diff
@@ -1930,6 +1930,594 @@ void NaturalGradientPerElementScaleComponent::Update(
+                                        const CuMatrixBase<BaseFloat> &out_deriv,
+                                        Component *to_update_in,
+                                        CuMatrixBase<BaseFloat> *in_deriv) const {
+  if (in_deriv)
+    static_cast<CuMatrix<BaseFloat>* >(in_deriv)->Resize(out_deriv.NumRows(), InputDim());
+  Convolutional1dComponent *to_update = dynamic_cast<Convolutional1dComponent*>(to_update_in);
+  int32 num_splice = InputDim() / patch_stride_;
+  int32 num_patches = 1 + (patch_stride_ - patch_dim_) / patch_step_;
+  int32 num_filters = filter_params_.NumRows();
+  int32 num_frames = out_deriv.NumRows();
+  int32 filter_dim = filter_params_.NumCols();
+
+  /** Buffer for backpropagation:
+   *  derivatives in the domain of 'patches_',
+   *  1row = vectorized rectangular feature patches,
+   *  1col = dim over speech frames,
+   */
+  CuMatrix<BaseFloat> patches_deriv(num_frames, filter_dim * num_patches, kSetZero);
+
+  //
+  // backpropagate to vector of matrices
+  // (corresponding to position of a filter)
+  //
+  std::vector<CuSubMatrix<BaseFloat>* > patch_deriv_batch, out_deriv_batch, 
+	  filter_params_batch;
+
+  CuSubMatrix<BaseFloat>* filter_params_elem = new CuSubMatrix<BaseFloat>(
+		  filter_params_, 0, filter_params_.NumRows(), 0, 
+		  filter_params_.NumCols());
+
+  // form batch in vector container
+  for (int32 p = 0; p < num_patches; p++) {
+    // form batch in vector container. for filter_params_batch, all elements 
+    // point to the same copy filter_params_elem
+    patch_deriv_batch.push_back(new CuSubMatrix<BaseFloat>(patches_deriv.ColRange(
+				    p * filter_dim, filter_dim)));
+    out_deriv_batch.push_back(new CuSubMatrix<BaseFloat>(out_deriv.ColRange(
+				    p * num_filters, num_filters)));
+    filter_params_batch.push_back(filter_params_elem);  
+  }
+  AddMatMatBatched(1.0f, patch_deriv_batch, out_deriv_batch, kNoTrans, 
+		  filter_params_batch, kNoTrans, 0.0f);
+
+  // release memory
+  delete filter_params_elem;
+  for (int32 p = 0; p < num_patches; p++) {
+    delete patch_deriv_batch[p];
+    delete out_deriv_batch[p];
+  }
+
+  // sum the derivatives into in_deriv
+  std::vector<int32> column_map(filter_dim * num_patches);
+  for (int32 p = 0, index = 0; p < num_patches; p++) {
+    for (int32 s = 0; s < num_splice; s++) {
+      for (int32 d = 0; d < patch_dim_; d++, index++) {
+        column_map[index] = p * patch_step_ + s * patch_stride_ + d;
+      }
+    }
+  }
+
+  if (in_deriv) {
+    std::vector<std::vector<int32> > reversed_column_map;
+    ReverseIndexes(column_map, InputDim(), &reversed_column_map);
+    std::vector<std::vector<int32> > rearranged_column_map;
+    RearrangeIndexes(reversed_column_map, &rearranged_column_map);
+    for (int32 p = 0; p < rearranged_column_map.size(); p++) {
+      CuArray<int32> cu_cols(rearranged_column_map[p]);
+      in_deriv->AddCols(patches_deriv, cu_cols);
```

Comment:
I think maybe this should be CopyCols.  Actually I'm very surprised that the tests passed because this component seems to add to its input and never zero it, yet the properties says that it just sets its input. TestSimpleComponentPropagateProperties() should have failed.  [see output_deriv2].